### PR TITLE
B1M2-656

### PR DESCRIPTION
**Pull request by prushin**

### CODE REVIEW

**Review by tim2CF — CHANGES_REQUESTED** [review]

File: package.yaml [hunk 1]
```diff
@@ -85,6 +94,7 @@ library:
   # Enable all warnings with -Weverything, then disable the ones we don’t care about
   - -Weverything
   - -Werror
+  - -Wno-missing-export-lists
```
[the remark this answers is not in the document]
Comment (prushin) [reply package.yaml:97]: Looks like not needed...

File: script/generate-protoc.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (prushin) [reply script/generate-protoc.sh:1]: done

File: package.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (prushin) [reply package.yaml:97]: Actually this needs to compile haskell files generated by protoc.

File: src/LndGrpc/Client.hs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    req ~ MethodInput s m,
+    res ~ MethodOutput s m
+  ) =>
+  ProtoLens.RPC s (m :: Symbol) ->
+  LndEnv ->
+  req ->
+  p (Either LndError res)
+runUnary rpc env req = do
+  res <- liftIO $ runClientIO $ do
+    grpc <- makeClient env True False
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndGrpc/Client.hs:33]: Tried to move it to env, looks like this client works for one request only.
So decided to think about move something to env later.

File: src/LndClient/Data/Type.hs [hunk 4]
```diff
@@ -8,20 +11,31 @@ where
...
+  | GrpcError G.ClientError
+  | LndGrpcError E.ClientError
   | GrpcUnexpectedResult Text
   | GrpcEmptyResult
   | LndError Text
   | LndEnvError Text
   | TChanTimeout Text
   deriving (Eq, Show)
 
+newtype FieldIndex
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/Type.hs]: Removed, thanks

**Review by tim2CF — CHANGES_REQUESTED** [review]

File: src/LndClient/RPC/TH.hs [hunk 5]
```diff
@@ -56,67 +59,48 @@ mkRpc k = do
...
-                InitWalletRequest
-                  { walletPassword =
-                      coerce $ envLndWalletPassword env,
-                    cipherSeedMnemonic =
-                      coerce seed,
-                    aezeedPassphrase =
-                      coerce $ envLndAezeedPassphrase env
-                  }
-          if isRight res
-            then waitForGrpc env
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/RPC/TH.hs]: yes, done

File: src/LndClient/Data/GetInfo.hs [hunk 6]
```diff
@@ -20,3 +25,14 @@ instance FromGrpc GetInfoResponse GRPC.GetInfoResponse where
...
       <*> fromGrpc (GRPC.getInfoResponseSyncedToGraph x)
+
+instance C2.FromGrpc GetInfoResponse LnGRPC.GetInfoResponse where
+  fromGrpc x =
+    GetInfoResponse
+      <$> mPubKey
+      <*> mSyncedToChain
+      <*> mSyncedToGraph
+    where
+      mSyncedToChain = maybeToRight (FromGrpcError "SyncedToChain is not set") $ x ^? LnGRPC.syncedToChain
```
Comment (tim2CF) [inline src/LndClient/Data/GetInfo.hs:36]: For scalar fields `^?` mightbe redundant, I think in most cases `^.` can be used + `maybe'*` lens for the messages which actually can be unset. But please verify it youself, it's just my assumptions

**Review by tim2CF — CHANGES_REQUESTED** [review]

File: src/LndClient/RPC/Generic.hs [hunk 7]
```diff
@@ -18,7 +21,13 @@ import qualified Control.Exception as CE
     catches,
     throw,
   )
+import Data.ProtoLens.Service.Types (HasMethod, HasMethodImpl (..))
+import GHC.TypeLits (Symbol)
+import GHC.TypeLits
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/RPC/Generic.hs:29]: Mistake, fixed

File: src/LndClient/RPC/Generic.hs [hunk 8]
```diff
@@ -190,6 +199,57 @@ grpcSubscribeSilent _ service method handler env req =
...
+    gB ~ MethodOutput s rm,
+    Show a,
+    Show b
+  ) =>
+  PL.RPC s (rm :: Symbol) ->
+  LndEnv ->
+  a ->
+  m (Either LndError b)
+grpcSync2Katip rpc env req =
+  katipAddContext (sl "RpcName" (show $ symbolVal rpc :: Text))
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/RPC/Generic.hs]: Done, thanks

File: src/LndClient/Data/GetInfo.hs (hunk 6 above)
Comment (prushin) [reply src/LndClient/Data/GetInfo.hs]: replaced it with ^.
for maybe fields maybe' lens can be used, so done

File: src/LndClient/RPC/TH.hs [hunk 9]
```diff
@@ -215,11 +206,10 @@ mkRpc k = do
       LndEnv ->
       SubscribeInvoicesRequest ->
       m (Either LndError ())
-    subscribeInvoices =
-      $(grpcSubscribe)
-        SubscribeInvoices
-        GRPC.lightningClient
-        GRPC.lightningSubscribeInvoices
+    subscribeInvoices handler =
```
Comment (tim2CF) [inline src/LndClient/RPC/TH.hs:209]: I think handler argument be eta-reduced

Comment (prushin) [reply src/LndClient/RPC/TH.hs]: yep, done

File: src/LndClient/Data/Invoice.hs [hunk 10]
```diff
@@ -33,6 +44,50 @@ instance FromGrpc Invoice GRPC.Invoice where
...
+      <*> invoiceValueMsat
+      <*> invoiceSettled
+      <*> invoiceSettleIndex
+      <*> invoiceMemo
+      <*> invoicePaymentRequest
+      <*> invoicePrivate
+      <*> invoiceAddIndex
+      <*> invoiceState
+    where
+      invoiceRHash = Right . RHash $ x ^. LnGRPC.rHash
```
Comment (tim2CF) [inline src/LndClient/Data/Invoice.hs:67]: I think hardcoding parser implementation to every line is a bad practice. Just use `fromGrpc` or `C2.fromGrpc`. Such simple hardcoded implementations might miss some important features like byte size verification, or some other kind of verification.

Comment (prushin) [reply src/LndClient/Data/Invoice.hs]: This was temporarily implementation, cleaned it up

File: src/LndClient/Data/NewAddress.hs [hunk 11]
```diff
@@ -16,3 +39,26 @@ instance FromGrpc NewAddressResponse GRPC.NewAddressResponse where
...
+    where
+      msg gAddrType gAccount =
+        defMessage
+          & LnGRPC.type' .~ gAddrType
+          & LnGRPC.account .~ gAccount
+
+instance C2.ToGrpc AddressType LnGRPC.AddressType where
+  toGrpc x =
+    case x of
+      WITNESS_PUBKEY_HASH -> Right LnGRPC.WITNESS_PUBKEY_HASH
```
Comment (drownbes) [inline src/LndClient/Data/NewAddress.hs:57]: right before case

File: src/LndClient/Data/Invoice.hs [hunk 12]
```diff
@@ -33,6 +44,39 @@ instance FromGrpc Invoice GRPC.Invoice where
...
+    case x of
+      LnGRPC.Invoice'OPEN -> Right OPEN
+      LnGRPC.Invoice'SETTLED -> Right SETTLED
+      LnGRPC.Invoice'CANCELED -> Right CANCELED
+      LnGRPC.Invoice'ACCEPTED -> Right ACCEPTED
+      _ -> Left $ FromGrpcError "InvoiceState value cannot be recognized"
+
+instance FromGrpc InvoiceState GRPC.Invoice_InvoiceState where
+  fromGrpc x =
+    case x of
```
Comment (drownbes) [inline src/LndClient/Data/Invoice.hs:78]: right before case

**Review by drownbes — CHANGES_REQUESTED** [review]

File: src/LndGrpc/Client.hs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  ProtoLens.RPC s (m :: Symbol) ->
+  LndEnv ->
+  req ->
+  (HeaderList -> res -> ClientIO ()) ->
+  p (Either LndError res)
+runStreamServer rpc env req handler = do
+  r <- liftIO $ runClientIO $ do
+    grpc <- makeClient env True False
+    rawStreamServer rpc grpc () req $ const handler
+  case r of
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndGrpc/Client.hs:58]: thanks

File: src/LndClient/RPC/TH.hs [hunk 14]
```diff
@@ -474,9 +422,15 @@ mkRpc k = do
...
-    grpcSync = case k of
-      RpcSilent -> [e|grpcSyncSilent|]
-      RpcKatip -> [e|grpcSyncKatip|]
-    grpcSubscribe = case k of
-      RpcSilent -> [e|grpcSubscribeSilent|]
-      RpcKatip -> [e|grpcSubscribeKatip|]
+    --grpcSync = case k of
+    --  RpcSilent -> [e|grpcSyncSilent|]
+    --  RpcKatip -> [e|grpcSyncKatip|]
+    grpcSync2 = case k of
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/RPC/TH.hs]: of course I cleaned up all temporary entities

File: src/LndClient/RPC/Generic.hs [hunk 15]
```diff
@@ -140,6 +152,117 @@ grpcSyncKatip rpcName service method env req =
...
+  PL.RPC s (rm :: Symbol) ->
+  (b -> IO ()) ->
+  LndEnv ->
+  a ->
+  m (Either LndError ())
+grpcSubscribe2Silent rpc handler env req =
+  case C2.toGrpc req of
+    Right grpcReq -> do
+      res <- runStreamServer rpc env grpcReq gHandler
+      case res of
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/RPC/Generic.hs:131]: yep

File: src/LndClient/RPC/Generic.hs [hunk 16]
```diff
@@ -140,6 +152,117 @@ grpcSyncKatip rpcName service method env req =
...
+    $ katipAddLndContext env
+    $ do
+      $(logTM) (newSev env InfoS) "RPC is running"
+      (ts, res) <-
+        liftIO $ stopwatch $
+          grpcSync2Silent rpc env req
+      katipAddContext (sl "ElapsedSeconds" (showElapsedSeconds ts)) $
+        case res of
+          Left e ->
+            katipAddContext (sl "RpcResponse" (show e :: Text)) $
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/RPC/Generic.hs]: done

File: src/LndClient/Data/NewAddress.hs (hunk 11 above)
Comment (prushin) [reply src/LndClient/Data/NewAddress.hs:51]: done

File: src/LndClient/Data/Invoice.hs (hunk 12 above)
Comment (prushin) [reply src/LndClient/Data/Invoice.hs]: done

Comment (prushin) [reply src/LndClient/Data/Invoice.hs]: there is one left at the bottom, so right cannot be out

File: src/LndClient/Data/HtlcEvent.hs [hunk 17]
```diff
@@ -13,16 +14,32 @@ data HtlcEvent
...
+instance FromGrpc HtlcEvent LnGRPC.HtlcEvent where
   fromGrpc x =
     HtlcEvent
-      <$> fromGrpc (GRPC.htlcEventIncomingChannelId x)
-      <*> fromGrpc (GRPC.htlcEventOutgoingChannelId x)
-      <*> fromGrpc (GRPC.htlcEventIncomingHtlcId x)
-      <*> fromGrpc (GRPC.htlcEventOutgoingHtlcId x)
-      <*> fromGrpc (GRPC.htlcEventTimestampNs x)
-      <*> fromGrpc (GRPC.htlcEventEventType x)
+      <$> fromGrpc (x ^. LnGRPC.incomingChannelId)
```
Comment (drownbes) [inline src/LndClient/Data/HtlcEvent.hs:31]: Its actually could be changed to fold. So you don't right fromGrpc all the time

File: src/LndClient/Data/Invoice.hs [hunk 18]
```diff
@@ -17,22 +19,36 @@ data Invoice
...
-      <*> fromGrpc (GRPC.invoiceSettled x)
-      <*> fromGrpc (GRPC.invoiceSettleIndex x)
-      <*> fromGrpc (GRPC.invoiceMemo x)
-      <*> fromGrpc (GRPC.invoicePaymentRequest x)
-      <*> fromGrpc (GRPC.invoicePrivate x)
-      <*> fromGrpc (GRPC.invoiceAddIndex x)
-      <*> first
-        (\e -> FromGrpcError $ "Invalid Invoice State" <> show e)
-        (enumerated $ GRPC.invoiceState x)
+      <$> fromGrpc (x ^. LnGRPC.rHash)
```
Comment (drownbes) [inline src/LndClient/Data/Invoice.hs:36]: same fold

File: src/LndClient/Data/ListInvoices.hs [hunk 19]
```diff
@@ -28,25 +31,24 @@ data ListInvoiceResponse
...
+          & LnGRPC.reversed .~ gReversed
 
-instance FromGrpc ListInvoiceResponse Grpc.ListInvoiceResponse where
+instance FromGrpc ListInvoiceResponse LnGRPC.ListInvoiceResponse where
   fromGrpc x =
     ListInvoiceResponse
-      <$> fromGrpc (Grpc.listInvoiceResponseInvoices x)
-      <*> fromGrpc (Grpc.listInvoiceResponseLastIndexOffset x)
-      <*> fromGrpc (Grpc.listInvoiceResponseFirstIndexOffset x)
+      <$> fromGrpc (x ^. LnGRPC.invoices)
```
Comment (drownbes) [inline src/LndClient/Data/ListInvoices.hs:52]: same fold

File: src/LndClient/Data/Newtype.hs [hunk 20]
```diff
@@ -126,21 +130,14 @@ instance FromGrpc (TxId a) Text where
...
-
 instance FromGrpc NodePubKey ByteString where
   fromGrpc = Right . NodePubKey
 
 instance FromGrpc NodePubKey Text where
   fromGrpc x =
     case B16.decode $ encodeUtf8 x of
       Right y -> Right $ NodePubKey y
-      Left {} -> Left $ ToGrpcError "NodePubKey hex decoding error"
+      Left {} -> Left $ FromGrpcError "NodePubKey hex decoding error"
```
Comment (drownbes) [inline src/LndClient/Data/Newtype.hs:140]: bimap?

File: src/LndClient/Data/OpenChannel.hs [hunk 21]
```diff
@@ -22,32 +29,80 @@ data OpenChannelRequest
...
+      LnGRPC.OpenStatusUpdate'ChanOpen cou -> OpenStatusUpdateChanOpen <$> fromGrpc cou
+      LnGRPC.OpenStatusUpdate'PsbtFund pf -> OpenStatusUpdatePsbtFund <$> fromGrpc pf
+
+instance FromGrpc ChannelOpenUpdate LnGRPC.ChannelOpenUpdate where
+  fromGrpc x = ChannelOpenUpdate <$> fromGrpc (x ^. LnGRPC.channelPoint)
+
+instance FromGrpc ReadyForPsbtFunding LnGRPC.ReadyForPsbtFunding where
+  fromGrpc x =
+    ReadyForPsbtFunding
+      <$> fromGrpc (x ^. LnGRPC.fundingAddress)
```
Comment (drownbes) [inline src/LndClient/Data/OpenChannel.hs:77]: fold

File: src/LndClient/Data/PayReq.hs [hunk 22]
```diff
@@ -14,9 +15,9 @@ data PayReq
...
   deriving (Eq, Show)
 
-instance FromGrpc PayReq GRPC.PayReq where
+instance FromGrpc PayReq LnGRPC.PayReq where
   fromGrpc x =
     PayReq
-      <$> fromGrpc (GRPC.payReqPaymentHash x)
-      <*> fromGrpc (GRPC.payReqNumMsat x)
-      <*> fromGrpc (GRPC.payReqExpiry x)
+      <$> fromGrpc (fromStrict $ x ^. LnGRPC.paymentHash)
```
Comment (drownbes) [inline src/LndClient/Data/PayReq.hs:21]: fold

**Review by drownbes — CHANGES_REQUESTED** [review]

File: src/LndClient/Data/HtlcEvent.hs (hunk 17 above)
Comment (prushin) [reply src/LndClient/Data/HtlcEvent.hs:31]: decided to postpone putting out fromGrpc

File: src/LndClient/Data/Newtype.hs (hunk 20 above)
Comment (prushin) [reply src/LndClient/Data/Newtype.hs]: yep

**Review by tim2CF — CHANGES_REQUESTED** [review]

File: nix/overlay/lnd.nix [hunk 23]
```diff
@@ -6,7 +6,7 @@
 
 buildGoModule rec {
   pname = "lnd";
-  version = "0.13.0-beta";
+  version = "0.13.1-beta.rc2";
```
[the remark this answers is not in the document]
Comment (prushin) [reply nix/overlay/lnd.nix:9]: done

File: src/LndClient/Data/Type.hs [hunk 24]
```diff
@@ -9,12 +12,14 @@ where
...
 import LndClient.Import.External
-import Network.GRPC.HighLevel.Generated
+--import Network.GRPC.HighLevel.Generated as G
+import Network.HTTP2.Client.Exceptions as E
 
 data LndError
   = ToGrpcError Text
   | FromGrpcError Text
-  | GrpcError ClientError
+  | -- | GrpcError G.ClientError
```
Comment (tim2CF) [inline src/LndClient/Data/Type.hs:21]: Redundant comment

File: src/LndClient/Orphan.hs [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+
+import qualified Data.Text.Internal as T
+import LndClient.Class
+import LndClient.Import.External
+
+instance FromGrpc Text T.Text where
+  fromGrpc = Right . fromStrict
+
+instance ToGrpc Text T.Text where
+  toGrpc = Right . toStrict
```
Comment (tim2CF) [inline src/LndClient/Orphan.hs:15]: It seems like proto-lens is using strict data? Then we don't need this file probably, we can use strict data in our own models

**Review by tim2CF — CHANGES_REQUESTED** [review]

File: src/LndClient/Data/Newtype.hs [hunk 26]
```diff
@@ -126,21 +130,12 @@ instance FromGrpc (TxId a) Text where
...
 instance FromGrpc NodePubKey ByteString where
   fromGrpc = Right . NodePubKey
 
 instance FromGrpc NodePubKey Text where
-  fromGrpc x =
-    case B16.decode $ encodeUtf8 x of
-      Right y -> Right $ NodePubKey y
-      Left {} -> Left $ ToGrpcError "NodePubKey hex decoding error"
+  fromGrpc =
+    bimap (const $ FromGrpcError "") NodePubKey . B16.decode . encodeUtf8
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/Newtype.hs:138]: yep, refactoring mistake

File: src/LndClient/Data/CloseChannel.hs [hunk 27]
```diff
@@ -66,28 +61,31 @@ instance ToGrpc CloseChannelRequest GRPC.CloseChannelRequest where
...
-          Close <$> fromGrpc a
-      GRPC.CloseStatusUpdate
-        (Just (GRPC.CloseStatusUpdateUpdateClosePending a)) ->
-          Pending <$> fromGrpc a
-      GRPC.CloseStatusUpdate Nothing ->
-        Right NothingUpdate
+instance FromGrpc CloseStatusUpdate LnGRPC.CloseStatusUpdate where
+  fromGrpc x = do
+    let update = x ^. LnGRPC.maybe'update
+    case update of
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/CloseChannel.hs:74]: yep

File: src/LndClient/Data/AddHodlInvoice.hs [hunk 28]
```diff
@@ -15,7 +20,13 @@ data AddHodlInvoiceRequest
       }
   deriving (Eq, Show)
 
-instance ToGrpc AddHodlInvoiceRequest GRPC.AddHoldInvoiceRequest where
+newtype AddHodlInvoiceResponse
+  = AddHodlInvoiceResponse
+      { paymentRequest :: PaymentRequest
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/AddHodlInvoice.hs]: Yes, removed

File: src/LndClient/Data/Channel.hs [hunk 29]
```diff
@@ -32,22 +34,22 @@ data PendingUpdate (a :: TxKind)
...
-      <$> fromGrpc (GRPC.channelRemotePubkey x)
-      <*> channelPointParser (GRPC.channelChannelPoint x)
-      <*> (toMSat <$> fromGrpc (GRPC.channelCapacity x))
-      <*> (toMSat <$> fromGrpc (GRPC.channelLocalBalance x))
-      <*> (toMSat <$> fromGrpc (GRPC.channelRemoteBalance x))
-      <*> (toMSat <$> fromGrpc (GRPC.channelCommitFee x))
-      <*> fromGrpc (GRPC.channelActive x)
+      <$> fromGrpc (fromStrict $ x ^. LnGRPC.remotePubkey)
+      <*> channelPointParser (fromStrict $ x ^. LnGRPC.channelPoint)
+      <*> (toMSat <$> fromGrpc (x ^. LnGRPC.capacity))
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/Channel.hs:42]: agree

File: src/LndClient/Data/Channel.hs [hunk 30]
```diff
@@ -32,22 +34,22 @@ data PendingUpdate (a :: TxKind)
...
   fromGrpc x =
     Channel
-      <$> fromGrpc (GRPC.channelRemotePubkey x)
-      <*> channelPointParser (GRPC.channelChannelPoint x)
-      <*> (toMSat <$> fromGrpc (GRPC.channelCapacity x))
-      <*> (toMSat <$> fromGrpc (GRPC.channelLocalBalance x))
-      <*> (toMSat <$> fromGrpc (GRPC.channelRemoteBalance x))
-      <*> (toMSat <$> fromGrpc (GRPC.channelCommitFee x))
-      <*> fromGrpc (GRPC.channelActive x)
+      <$> fromGrpc (fromStrict $ x ^. LnGRPC.remotePubkey)
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/Channel.hs:40]: Replaced all Lazy.Text with just Text

File: src/LndClient/Data/Type.hs (hunk 24 above)
Comment (prushin) [reply src/LndClient/Data/Type.hs]: yep

File: src/LndGrpc/Client.hs [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+  p (Either LndError res)
+runUnary rpc env req = do
+  res <- liftIO $ runClientIO $ do
+    grpc <- makeClient env True False
+    rawUnary rpc grpc req
+  return $ case res of
+    Right (Right (Right (_, _, (Right x)))) -> Right x
+    Left e -> Left $ LndGrpcError e
+    Right (Right (Right (_, _, (Left e)))) -> Left $ LndError $ pack e
+    _ -> Left $ LndError "LndGrpc request error"
```
Comment (tim2CF) [inline src/LndGrpc/Client.hs:40]: Maybe let's show actual error in text

File: src/LndGrpc/Client.hs [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+  (HeaderList -> res -> ClientIO ()) ->
+  p (Either LndError res)
+runStreamServer rpc env req handler = do
+  r <- liftIO $ runClientIO $ do
+    grpc <- makeClient env True False
+    rawStreamServer rpc grpc () req $ const handler
+  return $ case r of
+    Right (Right ((), _, _)) -> Right defMessage
+    Left e -> Left $ LndGrpcError e
+    _ -> Left $ LndError "LndGrpc request error"
```
Comment (tim2CF) [inline src/LndGrpc/Client.hs:60]: Maybe let's show actual error in text

**Review by tim2CF — CHANGES_REQUESTED** [review]

File: nix/overlay/ghc-source-gen.nix [hunk 33]
```diff
@@ -0,0 +1,22 @@
+{ mkDerivation, base, fetchgit, ghc, ghc-paths, hpack, QuickCheck
+, stdenv, tasty, tasty-hunit, tasty-quickcheck
+}:
+mkDerivation {
+  pname = "ghc-source-gen";
+  version = "0.4.0.0";
```
[the remark this answers is not in the document]
Comment (prushin) [reply nix/overlay/ghc-source-gen.nix]: done

File: nix/overlay/http2.nix [hunk 34]
```diff
@@ -0,0 +1,30 @@
+{ mkDerivation, aeson, aeson-pretty, array, base, base16-bytestring
+, bytestring, case-insensitive, containers, directory, doctest
+, filepath, gauge, Glob, heaps, hspec, http-types, mwc-random
+, network, network-byte-order, psqueues, stdenv, stm, text
+, time-manager, unordered-containers, vector
+}:
+mkDerivation {
+  pname = "http2";
+  version = "2.0.6";
```
[the remark this answers is not in the document]
Comment (prushin) [reply nix/overlay/http2.nix:9]: decided not to do it, as newer version of http2 will require to update bunch of other dependencies which may cause inconsistency

File: src/LndClient/Data/InitWallet.hs [hunk 35]
```diff
@@ -18,19 +22,16 @@ data InitWalletRequest
...
-          { GRPC.initWalletRequestWalletPassword =
-              gWalletPassword,
-            GRPC.initWalletRequestCipherSeedMnemonic =
-              gCipherSeedMnemonic,
-            GRPC.initWalletRequestAezeedPassphrase =
-              gAezeedPassphrase
-          }
+        defMessage
+          & LnGRPC.walletPassword .~ gWalletPassword
+          & LnGRPC.cipherSeedMnemonic .~ toList gCipherSeedMnemonic
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/InitWallet.hs]: yes

File: src/LndClient/Data/Newtype.hs [hunk 36]
```diff
@@ -105,7 +106,7 @@ instance ToGrpc NodePubKey ByteString where
 
 instance ToGrpc NodePubKey Text where
   toGrpc =
-    bimap (const $ ToGrpcError "UTF8_DECODE_ERROR") TL.fromStrict
+    bimap (const $ ToGrpcError "UTF8_DECODE_ERROR") id
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/Newtype.hs]: yep, put id temporarily, and just forgot to fix it

File: src/LndClient/Data/SubscribeChannelEvents.hs [hunk 37]
```diff
@@ -2,52 +2,65 @@
...
+  deriving (Eq, Ord, Show)
+
+instance FromGrpc UpdateType LnGRPC.ChannelEventUpdate'UpdateType where
+  fromGrpc x = case x of
+    LnGRPC.ChannelEventUpdate'OPEN_CHANNEL -> Right OPEN_CHANNEL
+    LnGRPC.ChannelEventUpdate'ACTIVE_CHANNEL -> Right ACTIVE_CHANNEL
+    LnGRPC.ChannelEventUpdate'INACTIVE_CHANNEL -> Right INACTIVE_CHANNEL
+    LnGRPC.ChannelEventUpdate'PENDING_OPEN_CHANNEL -> Right PENDING_OPEN_CHANNEL
+    LnGRPC.ChannelEventUpdate'CLOSED_CHANNEL -> Right CLOSED_CHANNEL
+    _ -> Left $ FromGrpcError "Wrong ChannelUpdateType"
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/SubscribeChannelEvents.hs]: done

File: src/LndGrpc/Client.hs (hunk 31 above)
Comment (prushin) [reply src/LndGrpc/Client.hs]: fixed

File: src/LndGrpc/Client.hs (hunk 32 above)
Comment (prushin) [reply src/LndGrpc/Client.hs]: fixed

File: src/LndClient/Data/Payment.hs [hunk 38]
```diff
@@ -4,23 +4,38 @@ module LndClient.Data.Payment
...
+      <*> fromGrpc (x ^. LnGRPC.status)
+
+instance FromGrpc PaymentStatus LnGRPC.Payment'PaymentStatus where
+  fromGrpc x =
+    case x of
+      LnGRPC.Payment'UNKNOWN -> Right UNKNOWN
+      LnGRPC.Payment'IN_FLIGHT -> Right IN_FLIGHT
+      LnGRPC.Payment'SUCCEEDED -> Right SUCCEEDED
+      LnGRPC.Payment'FAILED -> Right FAILED
+      _ -> Left $ FromGrpcError "Can't parse PaymentStatus"
```
[the remark this answers is not in the document]
Comment (prushin) [reply src/LndClient/Data/Payment.hs]: fixed

**Review by tim2CF — APPROVED** [review]
Good job!

**Pull request merged**